### DISCUSSION TITLE
Make NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE() available for out of line definition

### DISCUSSION THREAD

**Discussion by cyanidle** [nlohmann/json#4199] [Ideas]
Hello, I am making a code generator tool for defining JsonRPC types and methods

To improve compilation times there is an option to compile conversions from and to json out of line in a cpp file

The problem arises (at least on MSVC) due to NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE() having inline appended to the methods it generates:

![image](https://github.com/nlohmann/json/assets/96257000/8254d577-85a4-4b89-bf9e-afa90c623c73)

Proof:

1) Such definition does not cause link error 
![image](https://github.com/nlohmann/json/assets/96257000/f58175ef-f2e2-4c4b-84bf-356d6bfe26b5)

2) Adding *inline* causes link error
![image](https://github.com/nlohmann/json/assets/96257000/21fe9a3b-6b8f-4d0d-8955-fa76c02eba1d)

3) The solution could be a NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE() alternative NLOHMANN_DEFINE_TYPE_OUT_OF_LINE() without the *inline* keyword
![image](https://github.com/nlohmann/json/assets/96257000/ed5be649-9671-4939-b9e5-d71874b63101)

**Comment by cyanidle** [comment]
This is the header part of the type in question: 
![image](https://github.com/nlohmann/json/assets/96257000/883d728b-5227-469e-a1ce-5fed9d61e26e)

**Reply by gregmarr** [reply]
Given how many variations there already are for this macro, and I think more proposed, I would say it would be much simpler for you to have that macro in your code, or just have your code generator write this instead:

```
namespace common {
    void to_json(nlohmann::json& nlohmann_json_j, const Ack& nlohmann_json_t) { NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_TO, ok)) }
    void from_json(const nlohmann::json& nlohmann_json_j, Ack& nlohmann_json_t) { NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_FROM, ok)) }
}
```

I'm not the maintainer, just a user, so @nlohmann may have different ideas.

**Reply by cyanidle** [reply]
I did already patch the codegen. Even more macros probably is not the answer. But it was pretty convenient to use the macro in source (it does not break without namespaces for some reason) 

There could be a generic version of the original macro, which allows to place *friend* and *inline* parametrically, so that the two already present macros were made using one generic, but leave the option to customize.

If the maintainer is not against this idea, I could make a small PR with this change
